ng-dev/release/publish/actions.ts

```diff
@@ -563,50 +638,47 @@ export abstract class ReleaseAction {
+   *   additional changes after the release output has been built locally.
    * @param releaseNotes The release notes for the version being published.
    * @param publishBranch Name of the branch that contains the new version.
    * @param npmDistTag NPM dist tag where the version should be published to.
    * @param additionalOptions Additional options for building and publishing.
    */
-  protected async buildAndPublish(
+  protected async publish(
+    builtPackagesWithInfo: BuiltPackageWithInfo[],
     releaseNotes: ReleaseNotes,
+    beforeStagingSha: string,
     publishBranch: string,
     npmDistTag: NpmDistTag,
     additionalOptions: {skipExperimentalPackages?: boolean} = {},
   ) {
     const {skipExperimentalPackages} = additionalOptions;
-    const versionBumpCommitSha = await this._getCommitOfBranch(publishBranch);
+    const versionBumpCommitSha = await this.getLatestCommitOfBranch(publishBranch);
 
+    // Ensure the latest commit in the publish branch is the bump commit.
     if (!(await this._isCommitForVersionStaging(releaseNotes.version, versionBumpCommitSha))) {
       error(red(`  ✘   Latest commit in "${publishBranch}" branch is not a staging commit.`));
       error(red('      Please make sure the staging pull request has been merged.'));
       throw new FatalReleaseActionError();
     }
 
-    // Checkout the publish branch and build the release packages.
-    await this.checkoutUpstreamBranch(publishBranch);
-    // Install the project dependencies for the publish branch.
-    await this.installDependenciesForCurrentBranch();
-
-    // Note that we do not directly call the build packages function from the release
-    // config. We only want to build and publish packages that have been configured in the given
-    // publish branch. e.g. consider we publish patch version and a new package has been
-    // created in the `next` branch. The new package would not be part of the patch branch,
-    // so we cannot build and publish it.
-    const builtPackages = await invokeReleaseBuildCommand(this.projectDir);
-    const releaseInfo = await invokeReleaseInfoCommand(this.projectDir);
-
-    // Combine the built packages with their NPM package information. This is helpful
-    // later for filtering out e.g. experimental packages.
-    const builtPackagesWithInfo = mergeBuiltPackagesWithInfo(
-      builtPackages,
-      releaseInfo.npmPackages,
-    );
+    // Ensure no commits have landed since we started the staging process. This would signify
+    // that the locally-built release packages are not reflecting what is on GitHub.
+    if (!(await this._isRevisionAheadOfBase(beforeStagingSha, versionBumpCommitSha, 1))) {
+      error(red(`  ✘   Unexpected additional commits have landed while staging the release.`));
+      error(red('      Please revert the bump commit and retry, or cut a new version on top.'));
+      throw new FatalReleaseActionError();
```

Comment:
Nit: We could list the unexpected commit shas if you want to

Reply:
I agree, but it's probably not worth the restructuring/complications as this should happen rarely.